lib/Sema/CSGen.cpp

```diff
@@ -1937,40 +1937,23 @@ namespace {
 
     Type visitUnresolvedSpecializeExpr(UnresolvedSpecializeExpr *expr) {
       auto baseTy = CS.getType(expr->getSubExpr());
+      Type specializeTy = baseTy;
 
-      if (baseTy->isTypeVariableOrMember()) {
-        return baseTy;
+      if (AnyMetatypeType *meta = baseTy->getAs<AnyMetatypeType>()) {
+        specializeTy = meta->getInstanceType();
       }
 
-      // We currently only support explicit specialization of generic types.
-      // FIXME: We could support explicit function specialization.
-      auto &de = CS.getASTContext().Diags;
-      if (baseTy->is<AnyFunctionType>()) {
-        de.diagnose(expr->getSubExpr()->getLoc(),
-                    diag::cannot_explicitly_specialize_generic_function);
+      auto *overloadLocator = CS.getConstraintLocator(expr->getSubExpr());
+      if (addSpecializationConstraint(overloadLocator, specializeTy,
+                                      expr->getUnresolvedParams())) {
+        // Diagnosis already happened, but add a bit of context
+        auto &de = CS.getASTContext().Diags;
         de.diagnose(expr->getLAngleLoc(),
                     diag::while_parsing_as_left_angle_bracket);
         return Type();
```

Comment:
I'd prefer if we didn't diagnose and return `Type()` and subsequently fail the constraint generation because that means that the expression that contains this specialization won't get any code completion...

Reply:
In fact we should probably use `addUnsolvedConstraint` here to delay its simplification until solving begins.

Reply:
Now doesn't fail constraint generation, replaces missing types with `PlaceholderType`. I don't think there's any need to delay with `addUnsolvedConstraint`?